### ISSUE TITLE
FastArray subclasses requiring specific dtype should disallow dtype changes

### ISSUE THREAD

**Issue by OrestZborowski-SIG** [rtosholdings/riptable#211]
FastArray subclasses that expect a specific dtype (such as DateTimeNano which assumes i8) should disallow changing to a different dtype. This can happen explicitly with assignment, or implicitly in new-from-template via view().

_Originally posted by @wenjuno in https://github.com/rtosholdings/riptable/issues/207#issuecomment-914754615_

**Comment by OrestZborowski-SIG** [comment]
Note an interesting open issue in numpy regarding view-casting and inability to restrict dtype: https://github.com/numpy/numpy/issues/5669

**Comment by wenjuno** [comment]
I see, so the `__array_finalize__` doesn't know what's the dtype it's casting to. I'm actually thinking raising exception in `view` basically wastes this method and it could be better used by returning a FastArray with the new dtype, Otherwise it seems the only way I can get a integer array view of the DataTimeNano out is by using `asarray`, which is not particularly difficult. By the way, `astype` returns FastArray instead of DateTimeNano.

**Comment by OrestZborowski-SIG** [comment]
Correct. Since `view(dtype)` (what numpy calls "view casting") is expected to return the same type but with different dtype, then raising an exception for fixed-dtype classes like `DateTmeNano` seems the least-surprising option.

I'm not sure if returning a superclass (i.e. FastArray) in this case would be expected? It would technically change the meaning of the data from "time-based" to "generic-numeric".

You could explicitly use `view(type=FA, dtype)` to create a new instance (what numpy calls "new from template"), which will instantiate a different type using some dtype. Similarly, as you note the `astype()` method casts the type to a `FastArray`, reinterpreting the data.

**Comment by wenjuno** [comment]
Oh I didn't know view has a parameter for type. I guess that's another way to change the class other than `asarray`. I mean `astype` changes the class from `DateTimeNano` to `FastArray`, but it doesn't have the "type" parameter, so maybe it doesn't need to preserve the class.
